cw-orch-daemon/src/senders/no_sender.rs

```diff
@@ -1,63 +1,61 @@
-use crate::{error::DaemonError, tx_resp::CosmTxResponse, DaemonBase};
+use crate::{error::DaemonError, DaemonBase, GrpcChannel};
 
-use cosmrs::{AccountId, Any};
-use cosmwasm_std::Addr;
 use cw_orch_core::environment::ChainInfoOwned;
 
 use tonic::transport::Channel;
 
-use super::sender_trait::SenderTrait;
+use super::{builder::SenderBuilder, query::QuerySender};
 
-pub type QuerierDaemon = DaemonBase<NoSender>;
+/// Daemon that does not support signing.
+/// Will err on any attempt to sign a transaction or retrieve a sender address.
+pub type QueryOnlyDaemon = DaemonBase<NoSender>;
 
 /// Signer of the transactions and helper for address derivation
-/// This is the main interface for simulating and signing transactions
 #[derive(Clone)]
 pub struct NoSender {
     /// gRPC channel
-    pub grpc_channel: Channel,
+    pub channel: Channel,
     /// Information about the chain
     pub chain_info: ChainInfoOwned,
 }
 
-impl SenderTrait for NoSender {
+impl SenderBuilder for NoSender {
     type Error = DaemonError;
-    type SenderOptions = ();
-
-    async fn commit_tx_any(
-        &self,
-        _msgs: Vec<Any>,
-        _memo: Option<&str>,
-    ) -> Result<CosmTxResponse, DaemonError> {
-        unimplemented!("You used the DaemonQuerier, which can't send transactions");
-    }
+    type Options = ();
 
-    fn address(&self) -> Result<Addr, DaemonError> {
-        unimplemented!("You used the DaemonQuerier, which doesn't have an associated address");
-    }
+    async fn build(
+        chain_info: ChainInfoOwned,
+        _sender_options: Self::Options,
+    ) -> Result<Self, Self::Error> {
+        let channel = GrpcChannel::from_chain_info(&chain_info).await?;
 
-    fn msg_sender(&self) -> Result<AccountId, DaemonError> {
-        unimplemented!("You used the DaemonQuerier, which doesn't have an associated msg sender");
+        Ok(NoSender {
+            channel,
+            chain_info,
+        })
     }
+}
 
+impl QuerySender for NoSender {
     fn chain_info(&self) -> &ChainInfoOwned {
         &self.chain_info
     }
 
     fn grpc_channel(&self) -> Channel {
-        self.grpc_channel.clone()
+        self.channel.clone()
     }
 
-    fn set_options(&mut self, _options: Self::SenderOptions) {}
+    fn set_options(&mut self, _options: Self::Options) {}
+}
 
-    fn build(
-        chain_info: ChainInfoOwned,
-        grpc_channel: Channel,
-        _sender_options: Self::SenderOptions,
-    ) -> Result<Self, Self::Error> {
-        Ok(NoSender {
-            grpc_channel,
-            chain_info,
-        })
+#[cfg(test)]
+mod tests {
+    use cw_orch_networks::networks::JUNO_1;
+
+    use crate::DaemonBuilder;
+
+    #[test]
+    fn build() {
+        let builder: QueryOnlyDaemon = DaemonBuilder::new(JUNO_1).build_sender(())?;
```

Comment:
It's so clean :face_holding_back_tears:

Reply:
It does always require specifying the return type of the function because there's no way to know which SenderOptions turn into which Senders...

Do you think it's worth adding/refactoring some code to make that possible?

Reply:
> It does always require specifying the return type of the function because there's no way to know which SenderOptions turn into which Senders...
> 
> Do you think it's worth adding/refactoring some code to make that possible?

I think if we add PhantomData inside daemon builder to forward type to the constructed daemon and avoid specifying type, what do you think?

Reply:
Discussed removing `SenderOptions` type in favor of accepting `Sender` type itself in `fn sender_builder`.

Reply:
Actually this might not be that easy because the sender requires access to `ChainInfoOwned`, which is set in the daemon. 

I.e. if we remove senderoptions we could provide a sender that interacts with chain X while the daemon is configured for chain Y

Reply:
Changed it so `SenderBuilder` is now a trait that's implemented on `SenderOptions`, so providing options allows for determining Sender type